File: src/main/java/Dictionary.java
import mdag.MDAG;
import mdag.MDAGNode;

import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.List;
import java.util.stream.Collectors;

class Dictionary {

    private static MDAG dictionary;

    static void initialize() {
        Dictionary dictionary = new Dictionary();
        ClassLoader classLoader = dictionary.getClass().getClassLoader();
        InputStream inputStream = classLoader.getResourceAsStream("nsf2020.txt");
        List<String> words = new BufferedReader(new InputStreamReader(inputStream))
                .lines().collect(Collectors.toList());
        Dictionary.dictionary = (new MDAG(words));
    }

    static MDAG getDictionary() {
        return dictionary;
    }

    static MDAGNode getSourceNode() {
        return (MDAGNode) dictionary.getSourceNode();
    }
}

File: src/main/kotlin/Dictionary.kt
import Constants.VALID_LETTERS
import mdag.MDAG
import mdag.MDAGNode

object Dictionary {

    private val instance: MDAG

    init {
        val words = Main::class.java.getResourceAsStream("nsf2020.txt")
            .bufferedReader()
            .readLines()
            .filter(this::wordContainsValidLetters)
        instance = MDAG(words)
    }

    fun contains(word: String): Boolean {
        return instance.contains(word)
    }

    fun getSourceNode(): MDAGNode {
        return instance.sourceNode as MDAGNode
    }

    private fun wordContainsValidLetters(word: String): Boolean {
        word.forEach {
            if (!VALID_LETTERS.contains(it)) {
                return false
            }
        }
        return true
    }
}
